docs/docs/higher-order-components.md

```diff
@@ -0,0 +1,488 @@
+
+Components are the primary unit of code reuse in React. However, you'll find that some patterns aren't a straightforward fit for traditional components.
+
+For example, say you have a CommentList component that subscribes to an external data source to render a list of comments:
+
+```js
+class CommentList extends React.Component {
+  constructor(props) {
+    super();
+    this.handleChange = this.handleChange.bind(this);
+    this.state = {
+      // Store data in state
+      comments: DataSource.getComments()
+    };
+  }
+
+  componentDidMount() {
+    // Subscribe to changes
+    DataSource.addChangeListener(this.handleChange);
+  }
+
+  componentWillUnmount() {
+    // Clean up listener
+    DataSource.removeChangeListener(this.handleChange);
+  }
+
+  handleChange() {
+    // Update component state whenever the data source changes
+    this.setState({
+      comments: DataSource.getComments()
+    });
+  }
+
+  render() {
+    return (
+      <div>
+        {this.state.comments.map((comment) => (
+          <Comment comment={comment} key={comment.id} />
+        ))}
+      </div>
+    );
+  }}
+```
+
+Later, you write a component for subscribing to a single blog post, which follows a similar pattern:
+
+```js
+class BlogPost extends React.Component {
+  constructor(props) {
+    super();
+    this.handleChange = this.handleChange.bind(this);
+    this.state = {
+      blogPost: DataSource.getBlogPost(props.id)
+    };
+  }
+
+  componentDidMount() {
+    DataSource.addChangeListener(this.handleChange);
+  }
+
+  componentWillUnmount() {
+    DataSource.removeChangeListener(this.handleChange);
+  }
+
+  handleChange() {
+    this.setState({
+      blogPost: DataSource.getBlogPost(this.props.id)
+    });
+  }
+
+  render() {
+    return <BlogPost blogPost={this.state.blogPost} />;
+  }}
+```
+
+`CommentList` and `BlogPost` aren't identical—they call different methods on `DataSource`, and they render different output. But much of their implementation is the same:
+
+- On mount, add a change listener to `DataSource`.
+- Inside the listener, call `setState` whenever the data source changes.
+- On unmount, remove the change listener.
+
+You can imagine that in a large app, this same pattern of subscribing to `DataSource` and calling `setState` will occur over and over again. We want an abstraction that allows us to define this logic in a single place and share them across many components. This is where HOCs excel.
+
+We'll create a function called `withSubscription` that works like this:
+
+```js
+const CommentListWithSubscription = withSubscription(
+  // The first parameter is the wrapped component
+  CommentList,
+  // The second parameter retrieves the data we're interested in
+  (DataSource) => DataSource.getComments()
+);
+
+const BlogPostWithSubscription = withSubscription(
+  CommentList,
+  // An optional second argument passes the current props
```

Comment:
This comment doesn't match the similar comment a few lines above which makes it confusing. Are these two examples subtle different?

I'm not sold on the need for inline comments here at all. Maybe better to explain that after code? I found our ```javascript{lineNo,lineStart-lineEnd,...} notation super helpful for highlighting specific snippets and drawing attention to them.